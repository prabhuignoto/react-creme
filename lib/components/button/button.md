| props    | description                                                | datatype           | default |
| :------- | :--------------------------------------------------------- | :----------------- | :------ |
| children | content that will be embedded in the accordion body        | ReactNode          |         |
| disabled | Disables he button                                         | Boolean            |         |
| label    | Button label                                               | String             |         |
| onClick  | callback that will be called when a bread crumb is clicked | Function           |         |
| size     | Configures the size of the button                          | "sm" / "md" / "lg" | "sm"    |
| style    | CSS Style object                                           | Object             |         |