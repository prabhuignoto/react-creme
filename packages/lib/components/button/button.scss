@use '../../design/layout.scss';
@use '../../design/theme.scss';
@use '../../design/button.scss';
@use '../../design/animate.scss';
@use 'sass:map';

$types: (primary, default, danger);
$sizes: (sm, md, lg);

.rc-btn {
  &.rc-btn-no-border {
    border: 0;
    box-shadow: none;
  }

  &:not(.rc-btn-icon) {
    min-width: 50px;
  }
}

@each $value in $types {
  .rc-btn-#{$value} {
    @extend %button-#{$value};

    &.rc-disabled {
      @extend %disabled;
      border: 1px solid theme.$chinese-gray;
    }
  }
}

@each $size in $sizes {
  .rc-btn-#{$size} {
    @extend %text-#{$size};
    @extend %button-#{$size};

    .rc-btn-icon-container {
      @extend %icon-#{$size};
    }

    .rc-btn-label {
      @extend %font-#{$size};
    }
  }
}

.rc-btn-label {
  @extend %text-no-wrap;
  text-transform: capitalize;
}

@each $size in $sizes {
  .rc-btn-#{$size} {
    .rc-btn-label {
      @extend %font-#{$size};
    }

    .rc-btn-icon {
      @extend %icon-#{$size};
    }
  }
}

.rc-btn-icon {
  border-radius: 50%;
  padding: 0.25rem;

  &.rc-disabled {
    @extend %disabled;
  }

  &:not(.rc-disabled) {
    background: theme.$light-bg;
  }

  &:hover:not(.rc-disabled) {
    background: theme.$secondary;
  }
}

.rc-btn-icon-container + .rc-btn-label {
  margin-left: 0.25rem;
}

.rc-btn-progress-wrapper {
  margin-right: 0.5rem;
}
