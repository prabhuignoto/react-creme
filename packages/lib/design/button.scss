@use 'theme.scss';
@use 'core.scss';

%button {
  @extend %center;
  background: none;
  cursor: pointer;
  transition: background-color 50ms ease-in;
  white-space: nowrap;

  &:not(.dark) {
    @extend %border;
  }

  &.dark {
    @extend %border-radius;
  }

  svg {
    color: theme.$primary;
  }

  &:hover svg {
    color: theme.$white;
  }
}

@mixin get-button-type($bg, $bg-hover, $color, $color-hover) {
  background: $bg;
  color: $color;

  svg {
    color: $color;
  }

  &:hover {
    svg {
      color: $color-hover;
    }
  }

  &:hover:not(.rc-disabled) {
    background: $bg-hover;
    box-shadow: none;
    color: $color-hover;
  }
}

%button-primary {
  @include get-button-type(
    rgba(theme.$primary-rgb, 0.85),
    theme.$primary,
    theme.$white,
    theme.$white
  );
}

%button-danger {
  @include get-button-type(
    rgba(theme.$crimson, 0.9),
    theme.$crimson,
    theme.$white,
    theme.$white
  );
  border: 1px solid theme.$crimson;
}

%button-default {
  &.dark {
    @include get-button-type(
      theme.$cod-gray,
      theme.$primary,
      theme.$white,
      theme.$white
    );
  }

  &:not(.dark) {
    @extend %shadow-inset-small;
    @include get-button-type(
      theme.$white,
      theme.$primary,
      theme.$primary,
      theme.$white
    );
  }
}

%button-icon {
  @extend %button;
  border: 0;
}

%button-icon:hover {
  svg {
    color: theme.$white;
  }
}

@mixin get-button($size, $border) {
  @extend %button;

  &:not(.icon) {
    @if $size == small {
      height: 20px;
      padding: calc(20px * 0.25) calc(20px * 0.3);
    } @else if $size == large {
      height: 30px;
      padding: calc(30px * 0.25) calc(30px * 0.3);
    } @else if $size == medium {
      height: 25px;
      padding: calc(25px * 0.25) calc(25px * 0.3);
    }
  }

  &.icon {
    border: 0;
  }
}

%button-sm {
  @include get-button(small, true);
}

%button-md {
  @include get-button(medium, true);
}

%button-lg {
  @include get-button(large, true);
}

%button-sm-no-border {
  @include get-button(small, false);
}
%button-lg-no-border {
  @include get-button(large, false);
}
