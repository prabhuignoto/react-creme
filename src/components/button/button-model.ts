import React, { CSSProperties } from "react";

export interface ButtonModel {
  label?: string;
  onClick?: () => void;
  children?: React.ReactNode;
  disabled?: boolean;
  primary?: boolean;
  type?: "primary" | "default" | "danger" | "icon" | "progress";
  size?: "sm" | "md" | "lg";
  style?: CSSProperties;
  border?: boolean;
  focusable?: boolean;
}
