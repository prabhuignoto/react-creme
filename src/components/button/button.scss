@use "../../design/layout.scss";
@use "../../design/colors.scss";
@use "../../design/icon.scss";
@use "../../design/button.scss";
@use "../../design/font.scss";
@use "../../design/animate.scss";

$types: (primary, default, danger);
$sizes: (sm, md, lg);

.rc-btn {
  &:not(.rc-btn-danger) {
    @include animate.halo();
  }

  &.rc-btn-danger {
    @include animate.halo(4px, 4px);
  }

  &.rc-btn-no-border {
    border: 0;
    box-shadow: none;
  }
}

@each $value in $types {
  .rc-btn-#{$value} {
    @extend %button-#{$value};

    &.rc-disabled {
      @extend %disabled;
    }
  }
}

@each $size in $sizes {
  .rc-btn-#{$size} {
    @extend %font-#{$size};
    @extend %button-#{$size};

    .rc-btn-icon-container {
      @extend %icon-#{$size};
    }

    .rc-btn-label {
      @extend %font-#{$size};
    }
  }
}

.rc-btn-label {
  @extend %text-no-wrap;
  text-transform: capitalize;
}

.rc-btn-icon {
  border-radius: 50%;
  padding: 0.25rem;

  &.rc-disabled {
    @extend %disabled;
  }

  &:not(.rc-disabled) {
    background: colors.$light-bg;
  }

  &:hover:not(.rc-disabled) {
    background: colors.$secondary;
  }
}

.rc-btn-icon-container + .rc-btn-label {
  margin-left: 0.25rem;
}

.rc-btn-progress-wrapper {
  margin-right: 0.5rem;
}
