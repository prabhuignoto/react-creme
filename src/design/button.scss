@use "colors.scss";
@use "layout.scss";

%button {
  @extend %border;
  @extend %center;
  background: none;
  cursor: pointer;
  font-size: 1rem;
  height: 35px;
  margin: 0 0.25rem;
  padding: 0 0.5rem;
  transition: background-color 50ms ease-in;
  white-space: nowrap;

  svg {
    color: colors.$primary;
  }
}

@mixin get-button-type($bg, $bg-hover, $color, $color-hover) {
  background-color: $bg;
  color: $color;

  svg {
    color: $color;
  }

  &:hover {
    background-color: $bg-hover;
    color: $color-hover;

    svg {
      color: $color-hover;
    }
  }
}

%button-primary {
  @include get-button-type(
    colors.$primary,
    colors.$white,
    colors.$white,
    colors.$primary
  );
  border: 1px solid colors.$primary;
}

%button-danger {
  @include get-button-type(
    colors.$crimson,
    colors.$white,
    colors.$white,
    colors.$crimson
  );
  border: 1px solid colors.$crimson;
}

%button-default {
  @extend %shadow-inset;
  @include get-button-type(
    colors.$white,
    colors.$light-bg,
    colors.$black,
    colors.$black
  );
  border: 1px solid colors.$secondary;
}

%button-icon {
  @extend %button;
  border: 0;
  color: colors.$primary;
  // padding: 0;
}

%button-icon:hover {
  // background: colors.$primary;

  svg {
    color: colors.$tertiary;
  }
}

@mixin get-button($size, $border) {
  @extend %button;

  @if $size == small {
    svg {
      height: 70%;
      width: 70%;
    }
  } @else if $size == large {
    svg {
      height: 100%;
      width: 100%;
    }
  }
}

%button-sm {
  @include get-button(small, true);
}
%button-lg {
  @include get-button(large, true);
}

%button-sm-no-border {
  @include get-button(small, false);
}
%button-lg-no-border {
  @include get-button(large, false);
}
