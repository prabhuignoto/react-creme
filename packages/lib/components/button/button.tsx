import classNames from "classnames";
import React, { useMemo, useRef } from "react";
import { useFocus } from "../common/effects/useFocus";
import { CircularProgress } from "../progress/circular-progress";
import { ButtonModel } from "./button-model";
import "./button.scss";

const Button: React.FunctionComponent<ButtonModel> = ({
  border = true,
  children,
  disabled = false,
  focusable = true,
  type = "default",
  label = "",
  onClick,
  size = "sm",
  style = {},
}) => {
  // classes
  const buttonClass = useMemo(
    () =>
      classNames(
        {
          "rc-btn-default": type === "progress",
          "rc-btn-no-border": !border,
          "rc-disabled": disabled,
        },
        [`rc-btn-${size}`, `rc-btn-${type}`, "rc-btn"]
      ),
    [disabled]
  );

  // setup for focus

  const ref = useRef(null);

  if (focusable) {
    useFocus(ref, onClick);
  }

  const focusableProps = useMemo(
    () => ({
      tabIndex: focusable ? 0 : -1,
    }),
    []
  );

  // handler for button click
  const handleClick = () => !disabled && onClick?.();

  return (
    <button
      className={buttonClass}
      onClick={handleClick}
      ref={ref}
      role="button"
      style={style}
      {...focusableProps}
      aria-label={label}
    >
      {type === "progress" && !disabled && (
        <span className="rc-btn-progress-wrapper">
          <CircularProgress size={"xs"} />
        </span>
      )}
      {children && <span className="rc-btn-icon-container">{children}</span>}
      {label && type !== "icon" && (
        <span className="rc-btn-label">{label}</span>
      )}
    </button>
  );
};

export { Button };
