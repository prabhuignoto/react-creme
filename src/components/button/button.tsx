import classNames from "classnames";
import React, { useCallback, useMemo, useRef } from "react";
import "../../design/focus.scss";
import { useFocus } from "../common/effects/useFocus";
import { ButtonModel } from "./button-model";
import "./button.scss";

const Button: React.FunctionComponent<ButtonModel> = ({
  label,
  onClick,
  children,
  disabled,
  type = "default",
  size = "sm",
  style = {},
}) => {
  const buttonClass = useMemo(
    () =>
      classNames(
        {
          "rc-btn-disabled": disabled,
        },
        [`rc-btn-${size}`, `rc-btn-${type}`, "rc-btn"]
      ),
    [disabled]
  );

  const ref = useRef(null);

  useFocus(ref, { bgHighlight: false });

  const handleClick = useCallback(() => {
    if (!disabled) {
      onClick && onClick();
    }
  }, [disabled]);

  return (
    <button
      className={buttonClass}
      onClick={handleClick}
      ref={ref}
      role="button"
      tabIndex={0}
      style={style}
    >
      {children && <span className="rc-btn-icon-container">{children}</span>}
      {label && <span className="rc-btn-label">{label}</span>}
    </button>
  );
};

export { Button };
