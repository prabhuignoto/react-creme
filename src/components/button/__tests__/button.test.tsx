import { fireEvent, render } from "@testing-library/react";
import React from "react";
import { Button } from "../button";

const handler = jest.fn();

describe("Button", () => {
  it("should render default", () => {
    const { container } = render(<Button />);

    const button = container.firstChild;

    expect(button).toBeInTheDocument();
    expect(button).toHaveClass("button-default");
  });

  it("should render label", () => {
    const { getByText } = render(<Button label="My Button" />);
    expect(getByText("My Button")).toBeInTheDocument();
  });

  it("should render button snapshot", () => {
    const { getByRole } = render(<Button label="My Button" />);
    expect(getByRole("button")).toMatchSnapshot();
  });

  it("should call handler", () => {
    const { getByText } = render(
      <Button label="My Button" onClick={handler} />
    );

    fireEvent.click(getByText("My Button"));

    expect(handler).toBeCalled();
  });
});
