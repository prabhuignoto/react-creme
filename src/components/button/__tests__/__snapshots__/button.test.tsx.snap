// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should render button snapshot 1`] = `
<button
  class="button-default button-label-icon"
  role="button"
  tabindex="0"
>
  <span
    class="button-label"
  >
    My Button
  </span>
</button>
`;
